src/main/java/net/snowflake/ingest/streaming/internal/ParquetFlusher.java

```diff
@@ -0,0 +1,271 @@
+        writer.write(row);
+      }
+      writer.close();
+    } catch (Throwable t) {
+      logger.logError("Parquet Flusher: failed to write", t);
+      throw t;
+    }
+  }
+
+  public static class BdecParquetWriterBuilder
+      extends ParquetWriter.Builder<List<Object>, BdecParquetWriterBuilder> {
+    private final MessageType schema;
+    private final Map<String, String> extraMetaData;
+
+    protected BdecParquetWriterBuilder(
+        ByteArrayOutputStream stream, MessageType schema, Map<String, String> extraMetaData) {
+      super(new ByteArrayOutputFile(stream));
+      this.schema = schema;
+      this.extraMetaData = extraMetaData;
+    }
+
+    @Override
+    protected BdecParquetWriterBuilder self() {
+      return this;
+    }
+
+    @Override
+    protected WriteSupport<List<Object>> getWriteSupport(Configuration conf) {
+      return new BdecWriteSupport(schema, extraMetaData);
+    }
+  }
+
+  private static class ByteArrayOutputFile implements OutputFile {
+    private final ByteArrayOutputStream stream;
+
+    private ByteArrayOutputFile(ByteArrayOutputStream stream) {
+      this.stream = stream;
+    }
+
+    @Override
+    public PositionOutputStream create(long blockSizeHint) throws IOException {
+      stream.reset();
+      return new ByteArrayDelegatingPositionOutputStream(stream);
+    }
+
+    @Override
+    public PositionOutputStream createOrOverwrite(long blockSizeHint) throws IOException {
+      return create(blockSizeHint);
+    }
+
+    @Override
+    public boolean supportsBlockSize() {
+      return false;
+    }
+
+    @Override
+    public long defaultBlockSize() {
+      return (int) MAX_CHUNK_SIZE_IN_BYTES;
+    }
+  }
+
+  private static class ByteArrayDelegatingPositionOutputStream
+      extends DelegatingPositionOutputStream {
+    private final ByteArrayOutputStream stream;
+
+    public ByteArrayDelegatingPositionOutputStream(ByteArrayOutputStream stream) {
+      super(stream);
+      this.stream = stream;
+    }
+
+    @Override
+    public long getPos() {
+      return stream.size();
+    }
+  }
+
+  private static class BdecWriteSupport extends WriteSupport<List<Object>> {
+    MessageType schema;
+    RecordConsumer recordConsumer;
+    Map<String, String> extraMetaData;
+
+    // TODO: support specifying encodings and compression
+    BdecWriteSupport(MessageType schema, Map<String, String> extraMetaData) {
+      this.schema = schema;
+      this.extraMetaData = extraMetaData;
+    }
+
+    @Override
+    public WriteContext init(Configuration config) {
+      // TODO: metadata
+      return new WriteContext(schema, extraMetaData);
+    }
+
+    @Override
+    public void prepareForWrite(RecordConsumer recordConsumer) {
+      this.recordConsumer = recordConsumer;
+    }
+
+    @Override
+    public void write(List<Object> values) {
+      List<ColumnDescriptor> cols = schema.getColumns();
+      if (values.size() != cols.size()) {
+        throw new ParquetEncodingException(
```

Comment:
is this during write to the buffer? If yes, we will need to wrap it with SFException, because that's the only thing we will catch when OnErrorOption=CONTINUE

Reply:
this is during flushing and not really expected.

Reply:
do you mean to add `channelName` in the exception message?

Reply:
Yes, otherwise we wouldn't know which channel is having issue, right?